Modules/_typingmodule.c

```diff
@@ -27,9 +27,26 @@ _typing__idfunc(PyObject *module, PyObject *x)
     return x;
 }
 
+/*[clinic input]
+_typing.cast -> object
+
+    typ: object
+    val: object
+
+[clinic start generated code]*/
+
+static PyObject *
+_typing_cast_impl(PyObject *module, PyObject *typ, PyObject *val)
+/*[clinic end generated code: output=11224a3fa037a9a1 input=4dce97a747d81e0a]*/
+{
+    Py_INCREF(val);
```

Comment:
I have the same concern with @pablogsal  and I would like to request to typing modules authors to follow conventional CPython accelerated extension module structure which is written based on class, function unit not the partial instance method, etc.

Those extension modules maintain Python and C versions even though maintain cost exists.
And those decisions are based on that the accelerated gain is worth than maintain cost.
So If we feel that following conventional structure is too complicated as the maintenance view, 
IMHO, it means that the accelerated version is not worth writing.
For my example, I love to implement `vectorcall` but I decided to drop my `enumerate vectorcall` because the implementation was too complicated. (see https://github.com/python/cpython/pull/25154)

cc @Fidget-Spinner

Reply:
Anyway, this PR makes 2x faster on my local machine.

Mean +- std dev: [cast_base] 114 ns +- 5 ns -> [cast_pr] 56.0 ns +- 1.8 ns: 2.03x faster

`````

import pyperf

runner = pyperf.Runner()
runner.timeit(name='bench typing.cast',
              stmt='val = typing.cast(int, num)',
              setup = '''
import typing
import random
num = random.randint(0, 10000)
''')

`````

Reply:
@pablogsal @corona10  Thanks for your opinion, and I agree with your points.

Case of `typing.cast` is super simple like it was with `typing.NewType.__call__` and we can get x2 boost with small efforts and in my opinion it will be great to have such boost)

I understand your concerns regarding maintenance costs and it's valid point for me.

Reply:
> And those decisions are based on that the accelerated gain is worth than maintain cost.

@corona10 thank for pinging me. To be honest I don't know enough to comment. Properly benchmarking typing is very hard and still unsolved. IMO, there are 3 aspects to typing performance:

1. Static type checkers implemented in Python like mypy.
2. Runtime type checkers/introspection like Pydantic.
3. Runtime overhead of typed code vs fully untyped code.

All 3 will benefit from general CPython optimizations (eg. Vectorcall, specialization, cache). Although only 3. will benefit the most from speedups to typing module. 2. will benefit a little. 1. will have little benefit.

I'm working on covering these cases in pyperformance (I already submitted 1 open PR for case 1.). And I will send out an email to typing-sig for discussion in 2. and 3. soon.

To be clear, I am not against this PR at all (I'm +0). I just don't have enough data at the moment to decide :(.

Reply:
The entire reason _typing exists is that the identity function it houses was deemed improper to put inside _functools or _operator. I think the identity functions have a place as they are next to trivial development-wise but provide a significant performance improvement.

I understand @corona10's argument about providing a full symmetrical accelerated C module vs. a Python module but this currently neither necessary (because it's just two identity functions) nor particularly feasible (because it's a lot of duplicate work and typing is still evolving quite rapidly).

IMO it would be best just to take this improvement as is.

Reply:
> but this currently neither necessary (because it's just two identity functions) nor particularly feasible (because it's a lot of duplicate work and typing is still evolving quite rapidly).

That *is* necessary per PEP 0399

Reply:
In particular:

>This PEP requires that in these instances that the C code must pass the test suite used for the pure Python code to act as much as a drop-in replacement as reasonably possible